datafusion/core/src/physical_plan/file_format/parquet.rs

```diff
@@ -460,6 +498,20 @@ impl FileOpener for ParquetOpener {
     }
 }
 
+// Check PruningPredicates just work on one column.
+fn check_page_index_push_down_valid(predicate: &Option<PruningPredicate>) -> bool {
+    if let Some(predicate) = predicate {
+        // for now we only support pushDown on one col, because each col may have different page numbers, its hard to get
+        // `num_containers` from `PruningStatistics`.
+        let cols = predicate.need_input_columns_ids();
+        //Todo more specific rules
```

Comment:
Now, cause of `num_containers `  we only support only one col. In the future, we could add more specific rules.

Reply:
I don't think there is a guarantee that the data pages for each column break at the same row index -- since the statistics are stored per page (and thus for a single column) I am not sure when/if we'll be able to support multi column predicates (aka predicates that refer to different columns).

I tried to draw a diagram to illustrate what I think is going on:

```
┏━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ 
   ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ┃
┃     ┌──────────────┐  │     ┌──────────────┐  │  ┃
┃  │  │              │     │  │              │     ┃
┃     │              │  │     │     Page     │  │   
   │  │              │     │  │      3       │     ┃
┃     │              │  │     │   min: "A"   │  │  ┃
┃  │  │              │     │  │   max: "C"   │     ┃
┃     │     Page     │  │     │ first_row: 0 │  │   
   │  │      1       │     │  │              │     ┃
┃     │   min: 10    │  │     └──────────────┘  │  ┃
┃  │  │   max: 20    │     │  ┌──────────────┐     ┃
┃     │ first_row: 0 │  │     │              │  │   
   │  │              │     │  │     Page     │     ┃
┃     │              │  │     │      4       │  │  ┃
┃  │  │              │     │  │   min: "D"   │     ┃
┃     │              │  │     │   max: "G"   │  │   
   │  │              │     │  │first_row: 100│     ┃
┃     └──────────────┘  │     │              │  │  ┃
┃  │  ┌──────────────┐     │  │              │     ┃
┃     │              │  │     └──────────────┘  │   
   │  │     Page     │     │  ┌──────────────┐     ┃
┃     │      2       │  │     │              │  │  ┃
┃  │  │   min: 30    │     │  │     Page     │     ┃
┃     │   max: 40    │  │     │      5       │  │   
   │  │first_row: 200│     │  │   min: "H"   │     ┃
┃     │              │  │     │   max: "Z"   │  │  ┃
┃  │  │              │     │  │first_row: 250│     ┃
┃     └──────────────┘  │     │              │  │   
   │                       │  └──────────────┘     ┃
┃   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘  ┃
┃       ColumnChunk            ColumnChunk         ┃
┃            A                      B               
 ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━━ ━━┛
                                                    
                                                    
   Parquet RowGroup with 2 columns, A and B.        
                                                    
 Column A is stored in 2 pages and Column B is      
               stored in 3 pages                    
                                                    
                                                    
                                                   
```

                                                          
Given the predicate `A > 35 `, we could rule out page 1 using statistics  (max is 20)                                               
                                                          
This means we can avoid decoding row_index 0->199 for      ColumnChunk B (and skip page 3 entirely)                   
                                                          

## Multi Column Predicates                                                           
Given the predicate `A > 35 AND B = "F"`:                    
                                          
Using `A > 35`, we can rule out page 1 using statistics as before                                                     
                                                           
Using the `B = "F"` part, we could rule out pages 3 and 5 (row_index 0->99 and row_index 250->onward)                
                                                           
Thus only rows in indexes 100->244 need to be decoded  (Only those ranges are needed in the`RowSelector`)

Reply:
So among other things, this implies we probably want to break the pruning predicate down into conjuncts (aka split on `AND`) to try and isolate predicates that refer only to a single column

Reply:
It also means that the "number of containers" in Column A will be "2" (as it as pages 1 and 2) and the number of containers for Column B will be "3" (as it has pages 3, 4, 5).

Reply:
> So among other things, this implies we probably want to break the pruning predicate down into conjuncts (aka split on AND) to try and isolate predicates that refer only to a single column

Yes, i think it has been done in `FilterPushDownRule`, like break like rewrite the filter-expr to CNF( I means top level `operator` are all `and`).

But after check the code,  i am wrong there not any split logic in `logical plan`, i think we should do this in logical plan optimizer like in `filter pushdown` as early as possible and do it once🤔

Reply:
Maybe i can file an issue, to do this split on AND, in logical plan.

Reply:
I think this need a lot to think. here 
https://github.com/apache/arrow-datafusion/blob/096627a2340aa33526c46eb551283eb88daabbea/datafusion/core/src/physical_optimizer/pruning.rs#L445-L457
I think it checked the `refer only to a single column` 😂

Reply:
Forget above 😂, yes we need rewrite the pruning predicate down, combine by `and`, each sub-expr should refer only one col, if not we will not push it to parquet. 
after evaluate each sub-expr one by one will do the intersection get the final result.

The reason why we cannot evaluate muiti-column is in the `build_statistics_record_batch` it will return a recordBatch: each col must have the same length(not suitable for pageIndex).
https://github.com/apache/arrow-datafusion/blob/096627a2340aa33526c46eb551283eb88daabbea/datafusion/core/src/physical_optimizer/pruning.rs#L176-L177

Reply:
Yeah, I think this is somewhat more specialized than just breaking the expr down into CNF

The difference is that we can only evaluate conjuncts that have a single column reference